modules/rostests/kmtests/ntos_ps/PsQuota.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * PROJECT:     ReactOS kernel-mode tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Kernel mode tests for process quotas
+ * COPYRIGHT:   Copyright 2021 George Bișoc <george.bisoc@reactos.org>
+ */
+
+#include <kmt_test.h>
+
+static
+VOID
+QueryProcessQuotaLimits(VOID)
+{
+    NTSTATUS Status;
+    QUOTA_LIMITS QuotaLimits;
+
+    /* Query the quota limits */
+    Status = ZwQueryInformationProcess(NtCurrentProcess(),
+                                       ProcessQuotaLimits,
+                                       &QuotaLimits,
+                                       sizeof(QuotaLimits),
+                                       NULL);
+    if (skip(NT_SUCCESS(Status), "Failed to query quota limits -- %lx\n", Status))
+    {
+        return;
+    }
+
+    /* Report the limits */
+    trace("Process paged pool quota limit -- %lu\n", QuotaLimits.PagedPoolLimit);
+    trace("Process non paged pool quota limit -- %lu\n", QuotaLimits.NonPagedPoolLimit);
+    trace("Process page file quota limit -- %lu\n\n", QuotaLimits.PagefileLimit);
+}
+
+static
+VOID
+QueryProcessQuotaUsage(VOID)
+{
+    NTSTATUS Status;
+    VM_COUNTERS VmCounters;
+
+    /* Query the quota usage */
+    Status = ZwQueryInformationProcess(NtCurrentProcess(),
+                                       ProcessVmCounters,
+                                       &VmCounters,
+                                       sizeof(VmCounters),
+                                       NULL);
+    if (skip(NT_SUCCESS(Status), "Failed to query quota usage -- %lx\n", Status))
+    {
+        return;
+    }
+
+    /* Report the usage */
+    trace("Process paged pool quota usage -- %lu\n", VmCounters.QuotaPagedPoolUsage);
+    trace("Process paged pool quota peak -- %lu\n", VmCounters.QuotaPeakPagedPoolUsage);
+    trace("Process non paged pool quota usage -- %lu\n", VmCounters.QuotaNonPagedPoolUsage);
+    trace("Process non paged pool quota peak -- %lu\n", VmCounters.QuotaPeakNonPagedPoolUsage);
+    trace("Process page file quota usage -- %lu\n", VmCounters.PagefileUsage);
+    trace("Process page file quota peak -- %lu\n\n", VmCounters.PeakPagefileUsage);
+}
+
+START_TEST(PsQuota)
+{
+    NTSTATUS Status;
+    PEPROCESS Process = PsGetCurrentProcess();
+
+    /* Report the current process' quota limits */
+    QueryProcessQuotaLimits();
+
+    /* Charge some paged and non paged quotas */
+    Status = PsChargeProcessNonPagedPoolQuota(Process, 0x200);
+    ok_irql(PASSIVE_LEVEL);
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    Status = PsChargeProcessPagedPoolQuota(Process, 0x200);
+    ok_irql(PASSIVE_LEVEL);
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    /* Report the quota usage */
+    trace("=== QUOTA USAGE AFTER CHARGE ===\n\n");
+    QueryProcessQuotaUsage();
```

Comment:
Like you mean caching the quota values in local variables and then retry recharging/returning quotas again and check if the values have remained the same (with `ok()`)? Because on Windows the results are different for obvious reasons.

Reply:
Yeah, I would expect that
```
ZwQuery();
NonPagedUsage = VmCounters.QuotaNonPagedPoolUsage;

PsChargeProcessNonPagedPoolQuota(Process, 0x200);

ZwQuery();
ok(VmCounters.QuotaNonPagedPoolUsage == NonPagedUsage + 0x200);

PsReturnProcessNonPagedPoolQuota(Process, 0x200);

ZwQuery();
ok(VmCounters.QuotaNonPagedPoolUsage == NonPagedUsage);
```
would work, as long as no other code is running in your process (which it shouldn't, if you have APCs disabled)